angr/knowledge_base.py

```diff
@@ -1,30 +1,28 @@
-"""Representing the artifacts of a project."""
-
-from .knowledge_plugins.plugin import default_plugins
+from .errors import KnowledgeBaseNoPlugin
 
 
 class KnowledgeBase(object):
     """Represents a "model" of knowledge about an artifact.
 
-    Contains things like a CFG, data references, etc.
+    The knowledge base should contain as absolutely little redundant data
```

Comment:
Probably a more fundamental design question for @rhelmot or @ltfish but I was curious: could/should the KB ultimately eliminate the need for the original binary, a la IDA's IDB? I'm interested because of my work in #575. I'll eventually migrate that work to a "view" once that concept is worked out but should I rely on pulling from `elftools` and `pefile` at the time the function is run by the user or should that kind of information be extracted in an initial analysis run and either cached or stored in the KB?

Reply:
It is entirely possible to make the knowledge base a self-contained entity, that will allow reconstructing the binary based on the stored knowledge only. In that case the following additional information should be stored:

- All the data that from the ELF/PE header and program segments, for angr to be able to retrieve target architecture, platform, etc.
- Something like an 'externs' plugin, that contains addresses and names of SimProcedures that was created during the analysis, in order for certain indirect jumps targets to be meaningful.
- Maybe some other additional information.

In that case it should work, if I didn't miss something.

Reply:
Our original idea for the split was that the knowledgebase DB would hold all *facts* about a binary. Under that understanding, we could pull in the binary contents themselves, as those are facts. For saving and restoring, we'll need something like this to recreate the CLE loader object anyways (unless we want to always require the original binary, which is the case currently).

The compelling use-case (from IDA) is the ability to create and undo patches without needing to keep the original binary around.